src/main/webapp/app/core/util/alert.service.ts

```diff
@@ -1,38 +1,131 @@
+    onClose?: (alert: Alert) => void;
+    dismissible?: boolean;
+}
+
+export interface AlertCreationProperties extends AlertBase {
     translationKey?: string;
     translationParams?: { [key: string]: unknown };
-    timeout?: number;
-    toast?: boolean;
-    position?: string;
-    close?: (alerts: Alert[]) => void;
+    disableTranslation?: boolean;
 }
 
+interface AlertInternal extends AlertBase {
+    close: () => void;
+    isOpen: boolean;
+}
+
+export interface Alert extends Readonly<AlertInternal> {}
+
+const DEFAULT_TIMEOUT = 8000;
+const DEFAULT_DISMISSIBLE = true;
+
 @Injectable({
     providedIn: 'root',
 })
 export class AlertService {
-    timeout = 8000;
-    toast = false;
-    position = 'top right';
+    private alerts: AlertInternal[] = [];
 
-    // unique id for each alert. Starts from 0.
-    private alertId = 0;
-    private alerts: Alert[] = [];
+    errorListener: Subscription;
+    httpErrorListener: Subscription;
 
-    constructor(private sanitizer: DomSanitizer, private ngZone: NgZone, private translateService: TranslateService) {}
+    constructor(private sanitizer: DomSanitizer, private ngZone: NgZone, private translateService: TranslateService, private eventManager: EventManager) {
+        this.errorListener = eventManager.subscribe('artemisApp.error', (response: EventWithContent<unknown> | string) => {
+            const errorResponse = (response as EventWithContent<AlertError>).content;
+            this.addErrorAlert(errorResponse.message, errorResponse.translationKey, errorResponse.translationParams);
+        });
+
+        this.httpErrorListener = eventManager.subscribe('artemisApp.httpError', (response: any) => {
+            const httpErrorResponse = response.content;
+            switch (httpErrorResponse.status) {
+                // connection refused, server not reachable
+                case 0:
+                    this.addErrorAlert('Server not reachable', 'error.server.not.reachable');
+                    break;
+
+                case 400:
+                    if (httpErrorResponse.error !== null && httpErrorResponse.error.skipAlert) {
+                        break;
+                    }
+                    const arr = httpErrorResponse.headers.keys();
+                    let errorHeader = null;
+                    let entityKey = null;
+                    arr.forEach((entry: string) => {
+                        if (entry.toLowerCase().endsWith('app-error')) {
+                            errorHeader = httpErrorResponse.headers.get(entry);
+                        } else if (entry.toLowerCase().endsWith('app-params')) {
+                            entityKey = httpErrorResponse.headers.get(entry);
+                        }
+                    });
+                    if (errorHeader) {
+                        const entityName = translateService.instant('global.menu.entities.' + entityKey);
+                        this.addErrorAlert(errorHeader, errorHeader, { entityName });
```

Comment:
Context:
- Configure Grading page of a programming exercise
- trying to set a negative value for a test case weight

Two errors showed up:
- The first one from the error handling inside the component with the proper message. https://github.com/ls1intum/Artemis/blob/d1c506f74b7aa8da0ef5f8d3534e0d56a86a605b/src/main/webapp/app/exercises/programming/manage/grading/programming-exercise-configure-grading.component.ts#L357
- And an additional one from here with the content ‘error.settingNegative’.

The same effect can be seen wherever components define their own error handling for HttpErrors. The one mentioned above is the easiest I found to be reproduceable.

Reply:
Hmm. You're right, but I'm unsure as to what we can do about it. The second generic alert is generated by a HTTP interceptor which is in place for all requests. I don't think there's a way for it to know wether or not custom error handling exists.

We could drop the interceptor and require all requests to trigger their own alerts if necessary... 🤔

Reply:
As you say, the issue is that the interceptor intercepts before the custom error handling.

Probably it would be best to let the server return specific translation keys with optional key value pairs for a modified translation which then gets directly handled by the interceptor so there is no need for a custom handling within the client when something fails on the server.

Another possibility would be to register as the component an error handler and if the error occurs the error won't get passed to the eventManager but to the error handler. There could potentially be issues with deregistering the subscribers if the component closes but that would depend on the implementation.

It's not a small thing to do though so I would consider this out of scope but either you fix it immediately afterwards or please create an issue.

Reply:
I think most components implement their own error handling for HTTP errors. Therefore, I think it should be fine to disable the global error handler for now until a unified way to use the error translations that are currently spread all over the place is found.

Refactoring the error handling as such is definitely out of scope here.

Reply:
@julian-christl @b-fein So what are we doing? I could just remove this HTTP error handler block from the code. IMO, a component or service firing a request should catch errors on it and display appropriate alerts rather than having the server send translation keys - in the end, it should be the client code that is responsible for what's displayed in the UI. The server can still send details that can be filled in as translation parameters if needed in a specific use case

Reply:
I'll keep it like it is for now. We can bugfix this in a follow up if it happens to be a bigger problem.